Clean up compiler warnings in train-text

Davidy22 opened the pull request:
Running make made purple warning text appear on my screen and I like not seeing warnings when I run make

Some brackets to disambiguate order of operations and truncating some layer names by 1 more to avoid strncpy danger

Davidy22 committed:
Clean up compiler warnings in train-text

Some brackets to disambiguate order of operations

ggerganov commented on examples/train-text-from-scratch/train-text-from-scratch.cpp:
```diff
@@ -302,9 +302,9 @@ void init_model(struct my_llama_model * model) {
-        strncpy(layer.w3->name, (layers_i + ".feed_forward.w3.weight").c_str(), sizeof(layer.w3->name));
+        strncpy(layer.w1->name, (layers_i + ".feed_forward.w1.weight").c_str(), sizeof(layer.w1->name) - 1);
+        strncpy(layer.w2->name, (layers_i + ".feed_forward.w2.weight").c_str(), sizeof(layer.w2->name) - 1);
+        strncpy(layer.w3->name, (layers_i + ".feed_forward.w3.weight").c_str(), sizeof(layer.w3->name) - 1);
```
This change won't work - see the comment above the lines

slaren replied:
I would suggest using the recently added `ggml_format_name` here instead.

slaren replied:
If the name buffers are too small, let's increase the size instead of hacking non-null terminated strings.

Davidy22 replied:
I came back to report that I switched to `snprintf()` before committing and saw this. Won't need the string formatting capability but it looks like it saves characters

slaren commented on examples/train-text-from-scratch/train-text-from-scratch.cpp:
```diff
@@ -302,9 +302,9 @@ void init_model(struct my_llama_model * model) {
-        strncpy(layer.w1->name, (layers_i + ".feed_forward.w1.weight").c_str(), sizeof(layer.w1->name));
-        strncpy(layer.w2->name, (layers_i + ".feed_forward.w2.weight").c_str(), sizeof(layer.w2->name));
-        strncpy(layer.w3->name, (layers_i + ".feed_forward.w3.weight").c_str(), sizeof(layer.w3->name));
+        ggml_format_name(layer.w1, (layers_i + ".feed_forward.w1.weight").c_str());
```
The idea would be to do something like this:
```c++
gml_format_name(layer.w1, "%s.feed_forward.w1.weight", layers_i.c_str());
```
However, as @ggerganov said, this won't work due to the maximum size of 32 characters of the names.

I want to strongly reiterate that it is not ok to work around this by using a non-nul terminated string. We need to increase the max name length or otherwise fix it in a way that respects the buffer sizes.

Davidy22 replied:
I was going to suggest just trimming it down to feed_fwd, but some debug printouts tell me these names have to exactingly match names that I assume are in the model files

Seems like name length is determined by GGML_MAX_NAME, don't know what the consequences of fiddling that are. Bumping it up to any higher multiple of 2 should put it out of range of any reasonable potential name length. Changing it doesn't change anything, looking at output and places related to the name field it seems like there's quite a few places that need to be changed to change the effective name length, stripped this stuff out for now and I'll figure it out in the morning and it'll be a second commit.

slaren replied:
Increasing `GGML_MAX_NAME` should work. I would suggest increasing it by 16 so that the padding doesn't have to be changed.

Davidy22 replied:
Alright, I figured out why I thought more needed to be changed beyond `GGML_MAX_NAME`, turns out this change is breaking with models trained using the smaller name limit. Well, I don't know how much breakage it is exactly, the training script still seems to run when I drop the safety assert and feed in an old model file but if a similar safety check is being run anywhere else then bumping `GGML_MAX_NAME` is going to make model files that had previously truncated names stop. Well, I've added the commit to the pr. Currently this only cleans up a compiler warning about a sketchy string copy, if there's potential breakage maybe the name length bump can be left until a day comes when a necessary breaking change happens that the second commit can be folded in alongside.

Davidy22 committed:
Increase GGML_MAX_NAME

Avoiding strncpy danger in train-text-from-scratch and reducing potential future name length issues

slaren commented on examples/train-text-from-scratch/train-text-from-scratch.cpp:
```diff
@@ -294,17 +294,10 @@ void init_model(struct my_llama_model * model) {
+        ggml_format_name(layer.w2, "%s.feed_forward.w2.weight", layers_i.c_str());
+        ggml_format_name(layer.w3, "%s.feed_forward.w3.weight", layers_i.c_str());
+
         layer.w1->padding[0] = 0;
         layer.w2->padding[0] = 0;
         layer.w3->padding[0] = 0;
```
I am guessing that these were added as an attempt to use the padding as the nul terminator of the name, as initially the padding was right after the name in the `ggml_tensor` struct. As it turns out, the padding is no longer immediately after the name, so this wouldn't work anymore.

Anyway, this does nothing now and should be removed.

Davidy22 replied:
oh alright

Davidy22 replied:
oh alright
